### DISCUSSION TITLE
How to correctly Setup Airflow on M1

### DISCUSSION THREAD

**Discussion by dotsinspace** [apache/airflow#30148] [General]
I have tried many options from Docker-Compose to Dockerfile to pip. all of them failed except PIP. so i basically did `pip install apache-airflow` and airflow worked. Now what i am building is a simple Graphql Server which basically takes Edges ( From Reactflow ) and pass those to graphql server which runs following code.

`
import os
from airflow import DAG
from airflow.operators.python_operator import PythonOperator
from airflow.models import DagBag
from datetime import datetime
import inspect

os.environ['KMP_DUPLICATE_LIB_OK'] = 'True'

dag_directory = os.path.join(os.path.expanduser("~"), "airflow", "dags")

def my_python_function(**kwargs):
    print("Hello World")

def CreateDag(__edges, __title, __description):
    dag = DAG(
        dag_id=__title,
        description=__description,
        start_date=datetime.now(),
        default_view='tree',
        schedule_interval='@daily',
        catchup=False
    )
    create_operators(dag, __edges)
    sync_dag_to_db(dag)
    write_dag_to_file(dag)
    return dag

def write_dag_to_file(dag):
    # Save DAG to a file.
    dag_id = dag.dag_id
    file_name = f"{dag_id}.py"
    source_code = inspect.getsource(my_python_function)

    dag_directory = os.path.join(os.path.expanduser("~"), "airflow", "dags")
    with open(os.path.join(dag_directory, file_name), "w") as dag_file:
        dag_file.write(source_code)

def create_operators(dag, edges):
    operators = {}
    for item in edges:
        source = item['source']
        target = item['target']

        if source not in operators:
            operators[source] = PythonOperator(
                task_id=source,
                python_callable=my_python_function,
                provide_context=True,
                dag=dag,
                op_kwargs={'source': source}
            )

        if target not in operators:
            operators[target] = PythonOperator(
                task_id=target,
                python_callable=my_python_function,
                provide_context=True,
                dag=dag
            )

        operators[target].set_upstream(operators[source])

    return operators

def sync_dag_to_db(dag):
    dag.sync_to_db()
`

This code basically creates dag file in /Users/[username]/airflow/dags and Airflow even pickup those. but when i run triggers in airflow then it throws following error.

<img width="1840" alt="Screenshot 2023-03-16 at 10 01 42 PM" src="https://user-images.githubusercontent.com/76559517/225688557-46c80f73-1000-4eb3-8296-3c280d86b109.png">

I am new to this so kept everything simple and still it is not working. Please help me to fix it. and trust me on this from last 1 day i am trying CHAT-GPT and it failed to fix this issue.

Thanks and Regards

**Comment by potiuk** [comment]
I believe you are doing far too much . I am not sure why you are doing all the dag bag things. Just add your DAG in the dag folder, run scheduler (and keep it running), run webserver (and keep it running), and whatever other combination of other components you need and it should work. Scheduler will parse the files when it scans the dir and add the DAG for you. 

You are doing something completely strange, so no wonder it does not work.

**Reply by dotsinspace** [reply]
@potiuk i am trying following.

actually i have application based on reactflow which is basically dynamic user can add or delete node.

<img width="1840" alt="Screenshot 2023-03-16 at 10 20 44 PM" src="https://user-images.githubusercontent.com/76559517/225693528-4f6efed7-2cdc-4fbc-a092-09a6b1cdf595.png">

now i want to run these nodes as soon as root node gets triggered. i have already implemented Tree for this and used BFS to DFS for traversing all nodes and executing there functions it worked fine. but i want robust solution which can handle this and i encountered Apache airflow for this. everything is working fine but error show above kept coming.

**Reply by potiuk** [reply]
Because this is not how Airflow works. You cannot do it this way.

**Reply by dotsinspace** [reply]
can tell me any alternative to for this problem ( Instead of implementing Tree )

**Reply by dotsinspace** [reply]
i will be happy if you can share some thought on problem which i am trying to get solved. all suggestions are welcomed :)

**Reply by potiuk** [reply]
Sure. generate Python dag files, drop them into the DAG folder and wait until Scheduler processes them and creates the DAG. This is the way to go.

**Reply by dotsinspace** [reply]
was doing same when these error start poping up. now one more thing i am running docker-compose now but still no dags are showig when i checked dags file are there inside docker. then i checked .cfg file it is correct to but dags are not showing up.

**Reply by potiuk** [reply]
Can't help you with that sorry - if you are venturing on doing somethig that is completely not supported, you chose the difficult path and I am afraid you need to walk it alone.

**Reply by dotsinspace** [reply]
😭
